Content.Server/Discord/DiscordLink/DiscordLink.cs

```diff
@@ -0,0 +1,255 @@
+    public event Action<CommandReceivedEventArgs>? OnCommandReceived;
+    /// <summary>
+    ///     Event that is raised when a message is received from Discord. This is raised for every message, including commands.
+    /// </summary>
+    public event Action<SocketMessage>? OnMessageReceived;
+
+    public void RegisterCommandCallback(Action<CommandReceivedEventArgs> callback, string command)
+    {
+        OnCommandReceived += args =>
+        {
+            if (args.Command == command)
+                callback(args);
+        };
+    }
+
+    #endregion
+
+    public void Initialize()
+    {
+        _configuration.OnValueChanged(CCVars.DiscordGuildId, OnGuildIdChanged, true);
+        _configuration.OnValueChanged(CCVars.DiscordPrefix, OnPrefixChanged, true);
+        _client = new DiscordSocketClient(new DiscordSocketConfig()
+        {
+            GatewayIntents = GatewayIntents.All
+        });
+        _client.Log += Log;
+        _client.MessageReceived += OnCommandReceivedInternal;
+        _client.MessageReceived += OnMessageReceivedInternal;
+
+        if (_configuration.GetCVar(CCVars.DiscordToken) is not { } token || token == string.Empty)
+        {
+            _sawmill.Info("No Discord token specified, not connecting.");
+            // The Bot is not connected, so we need to set the client to null, because some methods check if the bot is connected using a null check on the client.
+            _client = null;
+            return;
+        }
+
+        // If the Guild ID is empty OR the prefix is empty, we don't want to connect to Discord.
+        if (_configuration.GetCVar(CCVars.DiscordGuildId) == string.Empty || _configuration.GetCVar(CCVars.DiscordPrefix) == string.Empty)
+        {
+            // This is a warning, not info, because it's a configuration error.
+            // It is valid to not have a Discord token set which is why the above check is an info.
+            // But if you have a token set, you should also have a guild ID and prefix set.
+            _sawmill.Warning("No Discord guild ID or prefix specified, not connecting.");
+            _client = null;
+            return;
+        }
+
+        _botToken = token;
+        // Since you cannot change the token while the server is running / the DiscordLink is initialized,
+        // we can just set the token without updating it every time the cvar changes.
+
+        _client.Ready += () =>
+        {
+            _sawmill.Info("Discord client ready.");
+            return Task.CompletedTask;
+        };
+
+        Task.Run(async () =>
+        {
+            try
+            {
+                await LoginAsync(token);
+            }
+            catch (Exception e)
+            {
+                _sawmill.Error("Failed to connect to Discord!", e);
+            }
+        });
+    }
+
+    public void Shutdown()
+    {
+        if (_client != null)
+        {
+            _sawmill.Info("Disconnecting from Discord.");
+            _client.Log -= Log;
```

Comment:
You probably shouldn't be unregistering this so that log messages during shutdown aren't eaten.